[RISCV] Remove `AND` mask generated by `( zext  ( atomic_load ) )` by replacing the load with `zextload` for orderings not stronger then monotonic.

janagor opened the pull request:
Extends changes from [ff687af](https://github.com/llvm/llvm-project/commit/ff687af04f5b0e85305250587b524cb0b3849aa0).
Fixes https://github.com/llvm/llvm-project/issues/131476.

This patch adds a DAG combine to replace an `AND` of an `ATOMIC_LOAD` with a full-bit mask (e.g. `0xFF`, `0xFFFF`, etc.) which is generated as a result of `(zext (atomic_load))`,  by a zero-extended load, provided the atomic operation is monotonic or weaker.

janagor committed:
Added pre-commit test.

janagor committed:
[RISCV] Optimized `and` with `atomic_load` into `zextload` when safe.

janagor committed:
Merge branch 'llvm:main' into riscv_remove_zext_from_bool_for_atomic_load

lenary commented:
The title of this PR is still wrong, it's not adding range metadata at all. It's doing some kind of combine. I don't know what the right name should be, maybe `[RISCV] Remove ZEXT after some atomic loads` or similar?

topperc commented:
I spoke with @preames and he told me that it is not safe to replace an ISD::ATOMIC_LOAD with an ISD::LOAD in the current SelectionDAG. I think there are optimizations that don't expect an ISD::LOAD to be atomic and make unsafe changes.

We should make an ISD::ATOMIC_LOAD here with ISD::ZEXTLOAD extension type. I made a couple patches yesterday to make that more possible 704fc6542cf171c603bd387663a2de3a3fa9c4ff and f6178cdad03468a9801a23f8f7e80927614fdb7e.

janagor committed:
Merge branch 'llvm:main' into riscv_remove_zext_from_bool_for_atomic_load

janagor committed:
Merge branch 'main' into riscv_remove_zext_from_bool_for_atomic_load

janagor commented:
For _testing purposes only_ I reverted the changes from this patch and made such change:
```
diff --git a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
index 53244a990a86..bb247574a5f2 100644
--- a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
@@ -386,6 +386,7 @@ SDValue DAGTypeLegalizer::PromoteIntRes_Atomic0(AtomicSDNode *N) {
     switch (TLI.getExtendForAtomicOps()) {
     case ISD::SIGN_EXTEND:
       ExtType = ISD::SEXTLOAD;
+      ExtType = ISD::ZEXTLOAD;
       break;
     case ISD::ZERO_EXTEND:
       ExtType = ISD::ZEXTLOAD;
```
just to check what is generated for a following snippet (the code from original [issue](https://github.com/llvm/llvm-project/issues/131476)):
```c
#include <stdbool.h>
extern bool t1;
bool test1(void) {
    return __atomic_load_n(&t1, __ATOMIC_RELAXED);
}
```
I run:
```bash
./bin/clang --target=riscv64 -O1 -S -emit-llvm atomic-ops-load.c -o test.ll
./bin/llc -filetype=asm test.ll -o test.s -O1 -mtriple=riscv64 --print-after-isel -debug 2> result.txt
```
I got following instructions:
```
test1:                                  # @test1
# %bb.0:                                # %entry
        lui     a0, %hi(t1)
        lb      a0, %lo(t1)(a0)
        ret
.Lfunc_end0:
```
with debug file info that I attached to this comment: [result.txt](https://github.com/user-attachments/files/19862687/result.txt). I think the main thing from it is the following info:
```
ISEL: Starting selection on root node: t9: i64,ch = AtomicLoad<(dereferenceable load monotonic (s8) from @t1), zext from i8> t0, t15
ISEL: Starting pattern match
  Initial Opcode index to 1152977
  TypeSwitch[i64] from 1152986 to 1152989
  Morphed node: t9: i64,ch = LB<Mem:(dereferenceable load monotonic (s8) from @t1)> t14, TargetGlobalAddress:i64<ptr @t1> 0 [TF=3], t0
ISEL: Match complete!
```

It seems to me that changing extention type of ISD::ATOMIC_LOAD is not enough. Does this imply that there should be a match for `(zext(atomic_load))` in tablegen that would catch this pattern?

topperc commented:
> For _testing purposes only_ I reverted the changes from this patch and made such change:
> 
> ```
> diff --git a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> index 53244a990a86..bb247574a5f2 100644
> --- a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> +++ b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> @@ -386,6 +386,7 @@ SDValue DAGTypeLegalizer::PromoteIntRes_Atomic0(AtomicSDNode *N) {
>      switch (TLI.getExtendForAtomicOps()) {
>      case ISD::SIGN_EXTEND:
>        ExtType = ISD::SEXTLOAD;
> +      ExtType = ISD::ZEXTLOAD;
>        break;
>      case ISD::ZERO_EXTEND:
>        ExtType = ISD::ZEXTLOAD;
> ```
> 
> just to check what is generated for a following snippet (the code from original [issue](https://github.com/llvm/llvm-project/issues/131476)):
> 
> ```c
> #include <stdbool.h>
> extern bool t1;
> bool test1(void) {
>     return __atomic_load_n(&t1, __ATOMIC_RELAXED);
> }
> ```
> 
> I run:
> 
> ```shell
> ./bin/clang --target=riscv64 -O1 -S -emit-llvm atomic-ops-load.c -o test.ll
> ./bin/llc -filetype=asm test.ll -o test.s -O1 -mtriple=riscv64 --print-after-isel -debug 2> result.txt
> ```
> 
> I got following instructions:
> 
> ```
> test1:                                  # @test1
> # %bb.0:                                # %entry
>         lui     a0, %hi(t1)
>         lb      a0, %lo(t1)(a0)
>         ret
> .Lfunc_end0:
> ```
> 
> with debug file info that I attached to this comment: [result.txt](https://github.com/user-attachments/files/19862687/result.txt). I think the main thing from it is the following info:
> 
> ```
> ISEL: Starting selection on root node: t9: i64,ch = AtomicLoad<(dereferenceable load monotonic (s8) from @t1), zext from i8> t0, t15
> ISEL: Starting pattern match
>   Initial Opcode index to 1152977
>   TypeSwitch[i64] from 1152986 to 1152989
>   Morphed node: t9: i64,ch = LB<Mem:(dereferenceable load monotonic (s8) from @t1)> t14, TargetGlobalAddress:i64<ptr @t1> 0 [TF=3], t0
> ISEL: Match complete!
> ```
> 
> It seems to me that changing extention type of ISD::ATOMIC_LOAD is not enough. Does this imply that there should be a match for `(zext(atomic_load))` in tablegen that would catch this pattern?

I just posted https://github.com/llvm/llvm-project/pull/137019 to check the extension type in isel patterns. Once that's in we can add new predicates for zextload.

topperc commented:
> > For _testing purposes only_ I reverted the changes from this patch and made such change:
> > ```
> > diff --git a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> > index 53244a990a86..bb247574a5f2 100644
> > --- a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> > +++ b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> > @@ -386,6 +386,7 @@ SDValue DAGTypeLegalizer::PromoteIntRes_Atomic0(AtomicSDNode *N) {
> >      switch (TLI.getExtendForAtomicOps()) {
> >      case ISD::SIGN_EXTEND:
> >        ExtType = ISD::SEXTLOAD;
> > +      ExtType = ISD::ZEXTLOAD;
> >        break;
> >      case ISD::ZERO_EXTEND:
> >        ExtType = ISD::ZEXTLOAD;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > just to check what is generated for a following snippet (the code from original [issue](https://github.com/llvm/llvm-project/issues/131476)):
> > ```c
> > #include <stdbool.h>
> > extern bool t1;
> > bool test1(void) {
> >     return __atomic_load_n(&t1, __ATOMIC_RELAXED);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I run:
> > ```shell
> > ./bin/clang --target=riscv64 -O1 -S -emit-llvm atomic-ops-load.c -o test.ll
> > ./bin/llc -filetype=asm test.ll -o test.s -O1 -mtriple=riscv64 --print-after-isel -debug 2> result.txt
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I got following instructions:
> > ```
> > test1:                                  # @test1
> > # %bb.0:                                # %entry
> >         lui     a0, %hi(t1)
> >         lb      a0, %lo(t1)(a0)
> >         ret
> > .Lfunc_end0:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with debug file info that I attached to this comment: [result.txt](https://github.com/user-attachments/files/19862687/result.txt). I think the main thing from it is the following info:
> > ```
> > ISEL: Starting selection on root node: t9: i64,ch = AtomicLoad<(dereferenceable load monotonic (s8) from @t1), zext from i8> t0, t15
> > ISEL: Starting pattern match
> >   Initial Opcode index to 1152977
> >   TypeSwitch[i64] from 1152986 to 1152989
> >   Morphed node: t9: i64,ch = LB<Mem:(dereferenceable load monotonic (s8) from @t1)> t14, TargetGlobalAddress:i64<ptr @t1> 0 [TF=3], t0
> > ISEL: Match complete!
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It seems to me that changing extention type of ISD::ATOMIC_LOAD is not enough. Does this imply that there should be a match for `(zext(atomic_load))` in tablegen that would catch this pattern?
> 
> I just posted #137019 to check the extension type in isel patterns. Once that's in we can add new predicates for zextload.

That patch landed. I have https://github.com/llvm/llvm-project/pull/137096 to try to fix the tablegen infrastructure. If that goes in we can use atomic_load_zext_8/16/32 for this patch.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -15267,6 +15267,48 @@ static SDValue reverseZExtICmpCombine(SDNode *N, SelectionDAG &DAG,
+  SDValue ZextLoad = DAG.getExtLoad(ISD::ZEXTLOAD, DL, ResultVT, Chain, Ptr,
+                                    MemOp->getPointerInfo(), LoadedVT,
+                                    MemOp->getAlign(), MemOp->getFlags());
+  DAG.ReplaceAllUsesOfValueWith(SDValue(N, 0), ZextLoad);
```
We need to make sure that the AND is the only user of the atomic_load data result. If another user depends on the sign extension behavior this will break.

topperc replied:
We should use `DCI.CombineTo(N, ZextLoad)` instead of `DAG.ReplaceAllUsesOfValueWith`

janagor committed:
Merge branch 'main' into riscv_remove_zext_from_bool_for_atomic_load

janagor committed:
fixup! [RISCV] Optimized `and` with `atomic_load` into `zextload` when safe.

janagor committed:
Merge branch 'riscv_remove_zext_from_bool_for_atomic_load' of github.com:janagor/llvm-project into riscv_remove_zext_from_bool_for_atomic_load

janagor committed:
Merge branch 'main' into riscv_remove_zext_from_bool_for_atomic_load

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -15274,6 +15274,50 @@ static SDValue reverseZExtICmpCombine(SDNode *N, SelectionDAG &DAG,
+  if (!MaskConst)
+    return SDValue();
+  uint64_t Mask = MaskConst->getZExtValue();
+  uint64_t ExpectedMask = LoadedVT.getSizeInBits() == 8    ? 0xFF
```
Can we use `maskTrailingOnes<uint64_t>(LoadedVT.getSizeInBits())`?

janagor committed:
Merge branch 'main' into riscv_remove_zext_from_bool_for_atomic_load

janagor committed:
fixup! [RISCV] Optimized `and` with `atomic_load` into `zextload` when safe.

used maskTrailingOnes

janagor committed:
Merge branch 'main' into riscv_remove_zext_from_bool_for_atomic_load

janagor committed:
fixup! [RISCV] Optimized `and` with `atomic_load` into `zextload` when safe.

Added pattern match for atomic_load_zext_n with replacement of getLoad with getAtomicLoad.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -15280,6 +15280,40 @@ static SDValue reverseZExtICmpCombine(SDNode *N, SelectionDAG &DAG,
+    return SDValue();
+
+  EVT LoadedVT = ALoad->getMemoryVT();
+  uint64_t Mask = maskTrailingOnes<uint64_t>(LoadedVT.getSizeInBits());
```
You're no longer getting the mask from the AND. Mask and ExpectedMask are always the same.

janagor committed:
fixup! fixup! [RISCV] Optimized `and` with `atomic_load` into `zextload` when safe.

corrected wrong usage of maskTrailingOnes

topperc reviewed (approved):
LGTM

janagor committed:
Merge branch 'main' into riscv_remove_zext_from_bool_for_atomic_load

janagor commented:
> LGTM

Thanks @topperc. I don’t have commit access, can you merge this PR for me?
